core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,413 @@
+            if (errors != Errors.NONE) {
+              responseCallback(initTransactionError(errors))
+            } else {
+              // init pid again
+              handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+            }
+          })
+      } else if (metadata.state == PrepareAbort || metadata.state == PrepareCommit) {
+        // wait for the commit to complete and then init pid again
+        txnMarkerPurgatory.tryCompleteElseWatch(new DelayedTxnMarker(metadata, (errors: Errors) => {
+          if (errors != Errors.NONE)
+            responseCallback(initTransactionError(errors))
+          else
+            handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+        }), Seq(metadata.pid))
+      } else {
+        metadata.producerEpoch = (metadata.producerEpoch + 1).toShort
+        metadata.txnTimeoutMs = transactionTimeoutMs
+        metadata.topicPartitions.clear()
+        metadata.lastUpdateTimestamp = time.milliseconds()
+        metadata.state = Empty
+        appendMetadataToLog(transactionalId, metadata, responseCallback)
+      }
+    }
+  }
+
+  private def validateTransactionalId(transactionalId: String): Errors =
+    if (transactionalId == null || transactionalId.isEmpty)
+      Errors.INVALID_REQUEST
+    else if (!txnManager.isCoordinatorFor(transactionalId))
+      Errors.NOT_COORDINATOR
+    else if (txnManager.isCoordinatorLoadingInProgress(transactionalId))
+      Errors.COORDINATOR_LOAD_IN_PROGRESS
+    else
+      Errors.NONE
+
+
+  def handleAddPartitionsToTransaction(transactionalId: String,
+                                       pid: Long,
+                                       epoch: Short,
+                                       partitions: collection.Set[TopicPartition],
+                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else {
+      // try to update the transaction metadata and append the updated metadata to txn log;
+      // if there is no such metadata treat it as invalid pid mapping error.
+      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+        case None =>
+          (Errors.INVALID_PID_MAPPING, null)
+
+        case Some(metadata) =>
+          // generate the new transaction metadata with added partitions
+          metadata synchronized {
+            if (metadata.pid != pid) {
+              (Errors.INVALID_PID_MAPPING, null)
+            } else if (metadata.producerEpoch != epoch) {
+              (Errors.INVALID_PRODUCER_EPOCH, null)
+            } else if (metadata.pendingState.isDefined) {
+              // return a retriable exception to let the client backoff and retry
+              (Errors.COORDINATOR_LOAD_IN_PROGRESS, null)
+            } else if (metadata.state != Empty && metadata.state != Ongoing) {
```

Comment:
@dguy `initPid` is only called only for the lifetime of a producer, when the producer client completed the current txn and is about to start the next one, it will not call `initPid` but just `addPartitions` as the first request. Only when producer client fails-over the new instance (i.e. in its next life) will send an `initPid` again.